### ISSUE TITLE
ZeroMQ shows errors in log "Redis::rPush(): send of 8192 bytes failed with errno=104"

### ISSUE THREAD

**Issue by chrisinmtown** [MISP/MISP#6771]
tl;dr - with the plugin ZeroMQ enabled we see recurring errors "Redis::rPush(): send of 8192 bytes failed with errno=104 Connection reset by peer" in multiple log files.  With setting Plugin.ZeroMQ_enable = False, no errors. 

### Work environment

| Questions                 | Answers
|---------------------------|--------------------
| Type of issue             | Support
| OS version (server)       | RedHat 7
| OS version (client)       | n/a
| PHP version               | 7.4
| MISP version / git hash   | 2.4.133
| Browser                   | n/a

### Support Questions

Breaking out this question from the (now closed) issue #4477 to try again for understanding and maybe resolution.  

Our RedHat 7 system runs MISP 2.4.133 and rh-redis5 (version 5.0.5).  We have enabled the zeromq plugin via Plugin.ZeroMQ_enable = true.  We see regular errors in log files as shown below.  Since these processes run on the same physical host (no network hop), the only thing I can guess might reset a connection from local PHP to local Redis is if the Redis server stops listening.  So far we cannot reproduce this on command. Here's the line of code that fails:

https://github.com/MISP/MISP/blob/9d537c989a1a9ff7de75bf5b4a0e1d72a92304a2/app/Lib/Tools/PubSubTool.php#L227

Unfortunately for us in version 2.4.133 there's no detailed PHP trace when this failure happens, so I don't know the exact ZeroMQ operation that's calling the pushToRedis() function in PubSubTool. I can wrap this in a simple PHP try-catch construct but I don't know how to gather more info in there - or how to enforce that a stack-trace is printed. 

In 2019, the advice posted to issue #4477 was (in part) to install php-redis ("yum install php-redis"). I'm struggling a bit with that advice in 2020 because I'm not at all certain the package ("php-pecl-redis.x86_64 0:2.2.8-1.el7") is a match for the PHP 7.4 that we are running in RedHat. Anyhow, I cannot find any directions in the installation documents about phpredis/php-redis, so maybe it's moot.

### Logs, screenshots, configuration dump, ...

debug.log:
```
2020-12-17 15:43:03 Notice: Redis::rPush(): send of 8192 bytes failed with errno=104 Connection reset by peer in [/var/www/MISP/app/Lib/Tools/PubSubTool.php, line 227]
2020-12-17 15:43:37 Notice: Redis::rPush(): send of 8192 bytes failed with errno=104 Connection reset by peer in [/var/www/MISP/app/Lib/Tools/PubSubTool.php, line 227]
```
resque-worker-error.log (the blank lines are in the log too):
```
Notice Error: Redis::rPush(): send of 8192 bytes failed with errno=104 Connection reset by
 peer in [/var/www/MISP/app/Lib/Tools/PubSubTool.php, line 227]

2020-12-17 15:43:03 Notice: Redis::rPush(): send of 8192 bytes failed with errno=104 Connection reset by peer in [/var/www/MISP/app/Lib/Tools/PubSubTool.php, line 227]
/var/www/MISP/app/tmp/logs/[58526]: 2020-12-17 15:43:04 Info: publish -- Event (180269): published.
Notice Error: Redis::rPush(): send of 8192 bytes failed with errno=104 Connection reset by peer in [/var/www/MISP/app/Lib/Tools/PubSubTool.php, line 227]

2020-12-17 15:43:37 Notice: Redis::rPush(): send of 8192 bytes failed with errno=104 Connection reset by peer in [/var/www/MISP/app/Lib/Tools/PubSubTool.php, line 227]
```

Here are the RPMs installed for PHP and redis; I don't see an obvious match for a "php-redis" package:
```
# rpm -qa | grep php
oniguruma5php-6.9.5+rev1-2.el7.remi.x86_64
php74-1.0-3.el7.remi.x86_64
php74-php-bcmath-7.4.7-1.el7.remi.x86_64
php74-php-cli-7.4.7-1.el7.remi.x86_64
php74-php-common-7.4.7-1.el7.remi.x86_64
php74-php-devel-7.4.7-1.el7.remi.x86_64
php74-php-fpm-7.4.7-1.el7.remi.x86_64
php74-php-json-7.4.7-1.el7.remi.x86_64
php74-php-mbstring-7.4.7-1.el7.remi.x86_64
php74-php-mysqlnd-7.4.7-1.el7.remi.x86_64
php74-php-opcache-7.4.7-1.el7.remi.x86_64
php74-php-pdo-7.4.7-1.el7.remi.x86_64
php74-php-pear-1.10.12-1.el7.remi.noarch
php74-php-pecl-zip-1.19.0-1.el7.remi.x86_64
php74-php-process-7.4.7-1.el7.remi.x86_64
php74-php-xml-7.4.7-1.el7.remi.x86_64
php74-runtime-1.0-3.el7.remi.x86_64
# rpm -qa | grep redis
python36u-redis-2.10.6-1.ius.centos7.noarch
rh-redis5-redis-5.0.5-1.el7.x86_64
rh-redis5-runtime-3.3-1.el7.x86_64
rh-redis5-3.3-1.el7.x86_64
```

**Comment by chrisinmtown** [comment]
I checked the size of our Redis store by connecting with redis-cli and using the 'info' command, here's the tail of that output:
```
# Keyspace
db0:keys=1144,expires=1112,avg_ttl=32387885
db13:keys=240190,expires=0,avg_ttl=0
```

**Comment by chrisinmtown** [comment]
Here's a bit of grasping at straws.  I see that the line of code that calls pubSubTool->publishEvent is guarded by a check on the ZeroMQ plugin:

https://github.com/MISP/MISP/blob/9d537c989a1a9ff7de75bf5b4a0e1d72a92304a2/app/Model/Event.php#L4691

I checked our configuration and found the plugin setting Plugin.ZeroMQ_enable =  True.  So maybe the failure lies in communicating with that pub-sub mechanism? In other words, MISP ingests and publishes an event just fine, tries to pass it on to the zeromq plugin, but when this redis error happens that event/attribute is not sent to mq subscribers as it should. It's difficult for a listener to know that it is missing events, and in our environment, we have only experimented with the zeromq listener, we don't rely on it.

**Comment by chrisinmtown** [comment]
I confirmed by running /var/www/MISP/tools/misp-zmq/sub.py that attribute-create events are being sent to the script via redis.

After digging a bit more, I believe but cannot yet prove that this error happens when the MISP server tries to publish an event JSON that's about 2.5MB to zeromq via redis.  I think the client is shoveling over blocks of data 8192 bytes at a time, then at some point the server stops accepting data ("broken connection"), so this error is logged.  But that size looks WELL within the bounds of what Redis can handle, so I'm baffled why the data does not flow cleanly.

Any ideas for debugging this further?  All suggestions welcome.

**Comment by chrisinmtown** [comment]
I'm pretty firmly convinced the error is due to ZeroMQ because the only place in the code I can find that calls rPush is in PubSubTool, and that's only called if Plugin.ZeroMQ_enable = True.
```
~/git/github/MISP% find . -type f | xargs grep '>rPush'   
./app/Lib/Tools/PubSubTool.php:        $redis->rPush($settings['redis_namespace'] . ':command', 'status');
./app/Lib/Tools/PubSubTool.php:            $redis->rPush($settings['redis_namespace'] . ':command', 'kill');
./app/Lib/Tools/PubSubTool.php:            $redis->rPush($settings['redis_namespace'] . ':command', 'reload');
./app/Lib/Tools/PubSubTool.php:                $redis->rPush($settings['redis_namespace'] . ':command', 'kill');
```

**Comment by chrisinmtown** [comment]
One last comment for today. I changed plugin setting Plugin.ZeroMQ_enable to False, restarted the MISP processes (PHP, Redis, workers, Apache), then waited two hours.  No Redis::rPush() errors anymore.  This settles it for me, the problem is somewhere linked to the ZeroMQ feature.
